phac-nml/irida-next
# Data Export - Email Notification

## What does this PR do and why?
This PR adds the email notification for users when their data export file is ready for download. Through the UI, users will have the option to select whether they want an email notification.

Note: Using template from Katherine's PR [519](https://github.com/phac-nml/irida-next/pull/519)

## Screenshots or screen recordings
Work Outlook client:
![image](https://github.com/phac-nml/irida-next/assets/82407232/f1f1f545-018f-4d4f-a623-1a719b688ea6)

Rails Preview:
![image](https://github.com/phac-nml/irida-next/assets/82407232/e6ce48f2-7380-4aae-babb-2a7facdcde0e)

Gmail:
![image](https://github.com/phac-nml/irida-next/assets/82407232/265bccbe-2182-4004-8b5a-f252b575d216)

Outlook in browser:
![image](https://github.com/phac-nml/irida-next/assets/82407232/c4a6cbe2-c43d-421d-b891-279aebac31d7)

Gmail Mobile app:
![image](https://github.com/phac-nml/irida-next/assets/82407232/13e2fd0e-5d6b-4f92-b042-caee18207361)

Outlook Mobile browser:
![image](https://github.com/phac-nml/irida-next/assets/82407232/d1e94980-2db0-4672-b86c-7314d2d71438)

With sans-serif font family:
Work Outlook client:
![image](https://github.com/phac-nml/irida-next/assets/82407232/e3a43424-d8fb-48ee-87f4-59025ed88110)

Gmail:
![image](https://github.com/phac-nml/irida-next/assets/82407232/cb4d0265-fb9a-4ae8-8d38-59e56ad0826a)

Outlook in browser:
![image](https://github.com/phac-nml/irida-next/assets/82407232/e4f59554-fb7c-4a51-8970-138f97b7bf7e)

Gmail mobile:
![image](https://github.com/phac-nml/irida-next/assets/82407232/f0c72525-7728-44f6-84e9-8d107f5b94a4)

Outlook mobile browser:
![image](https://github.com/phac-nml/irida-next/assets/82407232/79a3ec10-23fe-45f4-9307-9ca15d347ee4)

## How to set up and validate locally
Set up [Mailhog](https://github.com/mailhog/MailHog). Mailhog can be used to view emails without the need to send emails to real email address.
- In a terminal, run the go and mailhog install commands
- Run mailhog with command `~/go/bin/MailHog`
- Navigate to `http://0.0.0.0:8025/`. The mailhog email client should appear.

Preview email
- You can preview emails (like our components) by visiting `http://localhost:3000/rails/mailers/`
- For link clicking:
  - Seeded export 4 belongs to user3
  - Export ready without name belongs to admin user

Test sending email
- Restart your server if you haven't , otherwise Mailhog configs will not be set.
- In `rails c`, assign your params:
```
user = User.find_by(email: 'user1@email.com')
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Streptococcus agalactiae/Outbreak 2022 Sample 10').id]}, 'email_notification' => true, 'name' => 'Export name in test'}
```
- Run `DataExports::CreateService.new(user, params_1).execute`
- After 30 seconds, you should see the email appear in Mailhog
- Ensure the email appearance is similar to the screenshots above, with an expiry date of 3 business days from when you tested this, and export name `Export name in test`.

- Repeat the above with:
  - Removing `'name' => 'Export name in test'` from `params_1`. This should lead to an email with the `data_export.id` in place of `Export name in test`.
  - Removing `'email_notification' => true` from `params_1`. This should lead to no email notification after the `CreateService` has finished.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
